experimental/tpfand/tpfand.8
.\" Automatically generated by Pod::Man 2.23 (Pod::Simple 3.14)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "TPFAND 8"
.TH TPFAND 8 "2012-04-05" "Version 0.96.0" ""
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
tpfand \- Controls the fan speed of IBM/Lenovo ThinkPad\-Notebooks.
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
tpfand [\fB\-\-quiet\fR]
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBtpfand\fR controls the system fan of ThinkPad notebooks based on specified temperature profiles. Seperate trigger points can be configured for every temperature sensor in the notebook.
.IP "\fB\-\-quiet\fR" 8
.IX Item "--quiet"
Supresses startup messages
.SH "CONFIGURATION"
.IX Header "CONFIGURATION"
\&\fBtpfand\fR can either be configured interactively by running \fBtpfan-admin\fR or by modifying \fB/etc/tpfand.conf\fR. After manually changing the configuration file the daemon must be restarted for the new configuration to take effect.
.PP
You can specify the following options in \fB/etc/tpfand.conf\fR:
.IP "\fBenabled\fR = \fITrue / False\fR" 8
.IX Item "enabled = True / False"
If \fITrue\fR the system fan is controlled by \fBtpfand\fR, otherwise by the embedded hardware controller.
.IP "\fBoverride_profile\fR = \fITrue / False\fR" 8
.IX Item "override_profile = True / False"
If \fITrue\fR fan trigger temperatures and the options \fBhysteresis\fR, \fBinterval_speed\fR, \fBinterval_duration\fR and \fBinterval_delay\fR are read from the config file, otherwise from the profile matching your notebook stored in \fB/usr/share/tpfand/models\fR.
This option must be \fITrue\fR if you want to manually specify trigger temperatures.
.IP "\fBhysteresis\fR = \fIinteger\fR (K)" 8
.IX Item "hysteresis = integer (K)"
Specifies the hysteresis in K.
.IP "\fBinterval_speed\fR = \fIinteger\fR (1\-7)" 8
.IX Item "interval_speed = integer (1-7)"
Specifies the fan speed in interval cooling mode. Value must be between 1 (slowest) and 7 (fastest). Usually this should be set to 1.
.IP "\fBinterval_duration\fR = \fIinteger\fR (msec)" 8
.IX Item "interval_duration = integer (msec)"
Specifies the time in msec the fan may be running in interval cooling mode.
.IP "\fBinterval_delay\fR = \fIinteger\fR (msec)" 8
.IX Item "interval_delay = integer (msec)"
Specifies the minimum time in msec the fan must be off after running in interval colling mode.
.IP "\fBFan trigger temperature\fR" 8
.IX Item "Fan trigger temperature"
\&\fBSyntax:\fR [sensor id]. [sensor name] = [T1]:[F1] [T2]:[F2] ...
.Sp
where
.Sp
\&\fBsensor id\fR is the id of the sensor (0\-15).
.Sp
\&\fBsensor name\fR is a human readable sensor name, that will be shown by \fBtpfan-admin\fR.
.Sp
\&\fBTn\fR is a temperature in A\*^XC.
.Sp
\&\fBFn\fR is the fan level the system fan should run at when the temperature of this sensor rises about \fBTn\fR. 0 is off, 1 is interval cooling mode, 2\-8 are fan speeds.
.SH "BUGS"
.IX Header "BUGS"
Please report bugs at http://code.google.com/p/tpfanco/issues/list
.SH "AUTHORS"
.IX Header "AUTHORS"
\&\fBtpfand\fR was written by Sebastian Urban <surban84@googlemail.com> and is released under the \s-1GNU\s0 General Public License, version 3 or later.
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fBtpfan-admin\fR(1)


experimental/tpfand/src/tpfand.py
#! /usr/bin/python
# -*- coding: utf8 -*-
#
# tpfanco - controls the fan-speed of IBM/Lenovo ThinkPad Notebooks
# Copyright (C) 2011-2012 Vladyslav Shtabovenko
# Copyright (C) 2007-2009 Sebastian Urban
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#

import sys
if not ("/usr/share/pyshared" in sys.path):
    sys.path.append("/usr/share/pyshared")

import os
import os.path
import dbus
import dbus.service
from dbus.mainloop.glib import DBusGMainLoop
import signal
import gobject
from tpfand import build, settings, control, polkit

class Tpfand(object):

    quiet = False
    debug = False

    def __init__(self,quiet,debug):
        self.quiet = quiet
        self.debug = debug
        # register SIGTERM, SIGINT and sys.exit handlers
        signal.signal(signal.SIGTERM, self.term_handler)
        signal.signal(signal.SIGINT, self.term_handler)
        # register d-bus service
        gobject.threads_init()
        dbus.mainloop.glib.DBusGMainLoop(set_as_default=True)
        self.system_bus = dbus.SystemBus()
        bus_name = dbus.service.BusName(build.dbus_tpfand_name, self.system_bus)
        # initialize policy kit
        self.policy_kit = self.system_bus.get_object("org.freedesktop.PolicyKit1","/org/freedesktop/PolicyKit1/Authority")
        self.pk_authority = dbus.Interface(self.policy_kit, "org.freedesktop.PolicyKit1.Authority")
        self.tpfand_polkit = polkit.Polkit(self.system_bus,self.policy_kit,self.pk_authority)

        self.start_fan_control()
    def start_fan_control(self):
        """ daemon start function """
        if not self.quiet:
            print "tpfand " + build.version
            print "Copyright (C) 2011-2012 Vladyslav Shtabovenko"
            print "Copyright (C) 2007-2009 Sebastian Urban"
            print "This program comes with ABSOLUTELY NO WARRANTY"
            print
            print "WARNING: THIS PROGRAM MAY DAMAGE YOUR COMPUTER."
            print "         PROCEED ONLY IF YOU KNOW HOW TO MONITOR SYSTEM TEMPERATURE."
            print

        if self.debug:
            print "Running in debug mode"

        if not self.is_system_suitable():
            print "Fatal error: unable to set fanspeed or enable watchdog"
            print "             Please make sure you are root and a recent"
            print "             thinkpad_acpi module is loaded with fan_control=1"
            print "             If thinkpad_acpi is already loaded, check that"
            print "             " + build.ibm_fan + " exists. Thinkpad models"
            print "             that don't have this file are not supported."
            sys.exit(1)

        if os.path.isfile(build.pid_path):
            print "Fatal error: already running or " + build.pid_path + " left behind"
            sys.exit(1)

        # go into daemon mode
        self.daemonize()

    def daemonize(self):
        """ don't go into daemon mode if debug mode is active """
        if not debug:
            """go into daemon mode"""
            # from: http://aspn.activestate.com/ASPN/Cookbook/Python/Recipe/66012
            # do the UNIX double-fork magic, see Stevens' "Advanced
            # Programming in the UNIX Environment" for details (ISBN 0201563177)
            try:
                pid = os.fork()
                if pid > 0:
                    # exit first parent
                    sys.exit(0)
            except OSError, e:
                print >>sys.stderr, "fork #1 failed: %d (%s)" % (e.errno, e.strerror)
                sys.exit(1)

            # decouple from parent environment
            os.chdir("/")
            os.setsid()
            os.umask(0)

            # do second fork
            try:
                pid = os.fork()
                if pid > 0:
                    sys.exit(0)
            except OSError, e:
                print >>sys.stderr, "fork #2 failed: %d (%s)" % (e.errno, e.strerror)
                sys.exit(1)

        # write pid file
        try:
            pidfile = open(build.pid_path, "w")
            pidfile.write(str(os.getpid()) + "\n")
            pidfile.close()
        except IOError:
            print >>sys.stderr, "could not write pid-file: ", build.pid_path
            sys.exit(1)
        # start the daemon main loop
        self.daemon_main()

    def daemon_main(self):
        """daemon entry point"""
        # create and load configuration
        act_settings = settings.Settings(self.system_bus,"/Settings",self.tpfand_polkit,build,self.debug)

        # create controller
        self.controller = control.Control(self.system_bus,"/Control",act_settings,self.tpfand_polkit,build)

        #star glib main loop
        self.loop = gobject.MainLoop()
        self.loop.run()

    def is_system_suitable(self):
        """returns True if fan speed setting and watchdog are supported by kernel and
           we have write permissions"""
        try:
          fanfile = open(build.ibm_fan, "w")
          fanfile.write("level auto")
          fanfile.flush()
          fanfile.close()
          fanfile = open(build.ibm_fan, "w")
          fanfile.write("watchdog 5")
          fanfile.flush()
          fanfile.close()
          return True
        except IOError:
          return False

    def term_handler(self,signum, frame):
        """Handles SIGTERM and SIGINT"""
        self.controller.set_speed(255)
        print ""
        print "Terminating tpfand ..."
        try:
            os.remove(build.pid_path)
        except Exception, ex:
            #print "Can't delete " +  build.pid_path + ". Please do this by hand!"
            #print ex
            pass
        self.loop.quit()
        print "done!"

if __name__ == "__main__":
    quiet = False
    debug = False
    recreate_config = False
    if "--quiet" in sys.argv:
        quiet = True

    if "--debug" in sys.argv:
        debug = True

    app = Tpfand(quiet,debug)
